mgr/rook: Make use of rook-client-python when talking to Rook

sebastian-philipp opened the pull request:
~~**This PR is Blocked, until we have found a place for `rook-client-python` in the Rook github organization**~~

With this PR, I declare the `mgr/rook` as no longer a being prototype, as it finally fixes some major concerns:

* Files `rook_ceph_client` are automatically [generated](https://github.com/sebastian-philipp/rook-ceph-client-python). Thus the API can now be checked against the API definition. 
  The idea of `rook_ceph_client` to have them eventually in the Rook repository and not here. 
* We no longer craft json-patches by hand. Actually no one should ever do that.
  Also introduces a depencency to `python-jsonpatch`, as I no longer want to write those by hand.

Fixes issues related to the API not being in sync with reality:

* `CephFilesystem.spec.onlyManageDaemons` does not exist
* `CephObjectStroe.spec.gateway.allNodes` does not exist
* Added directory-osds to existsing nodes was broken

Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

* **Replaces #26028**
* Relates to https://github.com/rook/rook/issues/3543

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug
- [ ] Move https://github.com/sebastian-philipp/rook-ceph-client-python to the Ceph org

sebastian-philipp commented on src/pybind/mgr/rook/rook_cluster.py:
```diff
@@ -363,68 +364,60 @@ def add_osds(self, drive_group, all_hosts):
-                    "value": {'name': n}
-                })
-
-            new_dirs = list(set(directories) - set(current_node['directories']))
```
This should have been 
```
new_dirs = list(set(directories) -set([d['path'] for d in current_node['directories']]))
```

krig commented on src/pybind/mgr/rook/rook_cluster.py:
```diff
@@ -203,86 +209,81 @@ def add_filesystem(self, spec):
+            ),
+            spec=cnfs.Spec(
+                rados=cnfs.Rados(
+                    pool=spec.pool
```
Shouldn't this still be `spec.extended["pool"]`? Maybe I'm missing it, but I don't see how the spec object passed to `add_nfsgw` changed.

sebastian-philipp replied:
this will be changed in https://github.com/ceph/ceph/pull/29063/files#diff-45db35bc78bc3027f9421bd371b19e9bR239 , but that's not yet in master, right.

krig replied:
Ah, I see.

sebastian-philipp commented:
jenkins test make check

tchaikov commented on src/pybind/mgr/rook/rook_ceph_client/Readme.md:
```diff
@@ -0,0 +1,7 @@
+Files in here are automatically generated by 
```
what is this file for?

also i see we have couple "automatically generated" in this PR. is there anyway we can avoid adding them? or generate them at build time?

sebastian-philipp replied:
> also i see we have couple "automatically generated" in this PR. is there anyway we can avoid adding them? or generate them at build time?

Thing is, Rook's API is changing quite often and no one is fixing mgr/rook if the API changed. 
I'm open for suggestions of how to not fall behind changes in Rook's API. My idea ist:

1. [Add](https://github.com/rook/rook/pull/3534) JSON-schema validations to Rook's API
2. generate Python model classes out of those validations
3. Use Rook's API in a type-safe way so that we can verify we're *not* falling behind changes in Rook.

We're not going to catch behavior changes that are not reflected in the API definition, but still. 

---

Regarding not committing them. One idea would be to generate those at build time, but then we have to commit the schema definitions here?

sebastian-philipp replied:
@tchaikov another idea would be to make https://github.com/sebastian-philipp/rook-ceph-client-python a real project, build a package out of it and then depend on it?

tchaikov commented on debian/control:
```diff
@@ -278,6 +278,7 @@ Package: ceph-mgr-rook
 Architecture: all
 Depends: ceph-mgr (= ${binary:Version}),
          python-six,
+         python-jsonpatch
```
i think we should rely on `dh_python2` to do this for us. it is supposed to substitute `${python:Depends}` with the calculated dependencies. could you update the `override_dh_python2` target in `debian/rules` to run `dh_python2` against `ceph-mgr-rook` as well in a different commit and drop this change?

sebastian-philipp replied:
done

sebastian-philipp commented:
As a side effect, this fixes a TypeError:

```
debug 2019-09-18T14:11:55.310+0000 7f3ebcc2f700  0 mgr[rook] Completion Creating OSD on kubic-1:['vdb'] threw an exception:
Traceback (most recent call last):
  File "/usr/share/ceph/mgr/rook/module.py", line 189, in wait
    c.execute()
  File "/usr/share/ceph/mgr/rook/module.py", line 121, in execute
    self._result = self.execute_cb()
  File "/usr/share/ceph/mgr/rook/module.py", line 457, in execute
    return self.rook_cluster.add_osds(drive_group, all_hosts)
  File "/usr/share/ceph/mgr/rook/rook_cluster.py", line 536, in add_osds
    new_dirs = list(set(directories) - set(current_node['directories']))
TypeError: 'NoneType' object is not iterable
```

sebastian-philipp commented:
Hm. Starting to believe that having the generated model classes in Ceph doesn't make sense. There would be a constant stream of PRs just for pulling the latest changes.

sebastian-philipp commented:
@tchaikov @smithfarm what do you think of https://github.com/ceph/ceph/pull/29427/commits/45bd0046f9db1fb61e3da0ed376eaa7ab5ace537 ?

- [ ] Add make-dist integration

smithfarm commented:
@sebastian-philipp Fine with me, assuming it works?

smithfarm commented on src/pybind/mgr/rook/generate_rook_ceph_client.sh:
```diff
@@ -0,0 +1,10 @@
+. .venv/bin/activate
+pip install -e git+https://github.com/sebastian-philipp/rook-ceph-client-python.git#egg=rook-ceph-client
+
+curl https://raw.githubusercontent.com/rook/rook/master/cluster/examples/kubernetes/ceph/common.yaml | generate-model-classes --create-package  - rook_ceph_client
```
In my experience, GitHub is not 100% reliable, so I assume this line will fail from time to time.

tchaikov commented:
@sebastian-philipp TBH, i don't know. it's a dilemma. on one hand, i don't like put generated files in repo, on the other hand, i don't want to pull in personal repo as build dependencies of ceph/ceph. i understand there is a gray area w.r.t. "personal repo", as we use PPA and other sources / mirrors / registry when building Ceph and the document. but when it comes to building ceph/ceph, i am inclined to 

1. use repo under the ceph organization.
2. ensure the build is repeatable
3. have a self-contained dist tarball

do this make sense to you?

sebastian-philipp commented:
> do this make sense to you?

Yes, totally. 

> @sebastian-philipp TBH, i don't know. it's a dilemma. 

Indeed. 

> on one hand, i don't like putting generated files in repo, on the other hand, i don't want to pull in personal repo as build dependencies of ceph/ceph. i understand there is a gray area w.r.t. "personal repo", as we use PPA and other sources / mirrors / registry when building Ceph and the document. but when it comes to building ceph/ceph, i am inclined to
> 
> 1. use repo under the ceph organization.

I'd be really happy, if this would move completely out of my profile's scope! Do you think this would be possible?

> 2. ensure the build is repeatable

If we specify the exact dependencies to the common.yaml and the generator, would that be ok?

> 3. have a self-contained dist tarball

Adding this to make-dist is next on the list for this PR. Similar to the dashboard.

I'm happy that those things are actually solvable.

smithfarm commented:
On the face of it, the make-dist changes look good to me. (I'm assuming the cmake part of this PR will take the files this change is adding to the tarball and install them in the place where the spec file will find them, and assign to the right subpackage. I have not checked that part.)

sebastian-philipp commented:
> On the face of it, the make-dist changes look good to me. (I'm assuming the cmake part of this PR will take the files this change is adding to the tarball and install them in the place where the spec file will find them, and assign to the right subpackage. I have not checked that part.)

Yeah, `ceph-mgr-rook` will just package the content of this directory. 

But somehow this whole things feels wrong to me. https://github.com/sebastian-philipp/rook-ceph-client-python is really close before it is usable as a stand-alone Python pip package. Not making us of it feels odd.

smithfarm commented:
> a stand-alone Python pip package. Not making us of it feels odd.

Standalone pip package is not sufficient. It would need to be packaged for all the supported distros. Then it could be added to the RPM spec and `debian/control` files as a runtime dependency of ceph-mgr-rook.

sebastian-philipp commented:
> This PR could be seen as a temporary workaround until all that packaging work is completed.

fine with me.

Let's see if I can get this into the rook org first: https://github.com/rook/rook/issues/3543#issuecomment-536439823

tserong commented:
>  Then it could be added to the RPM spec and `debian/control` files as a runtime dependency of ceph-mgr-rook (if that's appropriate) (?)

If we do ultimately get an RPM of rook-ceph-client-python, this *still* involves running `curl https://raw.githubusercontent.com/rook/rook/master/cluster/examples/kubernetes/ceph/common.yaml` at some point, right? If that's the case, then this still needs to be inside `make dist`, rather than being a runtime thing.

sebastian-philipp commented:
unstale.

sebastian-philipp commented:
rebased

jmolmo commented:
I think that to use the Rook client python classes make the work more easy and the code more understandable and even beauty. 
But i see several problems (most of them commented in this PR) until we install this code as an external dependency.
I think that is not good to generate the code automatically,  we should use a fixed version of the Rook Client library, tested against a  fixed set of Rook versions. IMHO i just would leave a copy of the generated python files in the folder "rook_ceph_client" until we have the possibility of manage this as an external python dependence. 
I also would remove everything linked/related with the automatic generation form the ceph repo. 
I think that this will make live more easy to people that take care of the repo and the builds, and people that is interested in work with this nice Python classes.

sebastian-philipp commented:
> I think that is not good to generate the code automatically, we should use a fixed version of the Rook Client library, tested against a fixed set of Rook versions. IMHO i just would leave a copy of the generated python files in the folder "rook_ceph_client" until we have the possibility of manage this as an external python dependence.

Fixed! The code is now static.
> I also would remove everything linked/related with the automatic generation form the ceph repo.
Fixed! We're now just making a `pip install` to fetch the files.
> I think that this will make live more easy to people that take care of the repo and the builds, and people that is interested in work with this nice Python classes.

**Changes**:

* With [rook/rook3404](https://github.com/rook/rook/pull/4204#issuecomment-579485806) now being in limbo for the foreseeable future, Let's use my personal git repo for now. 
* I've removed all code generation here. Now, we only do a `pip install` and add the files to mgr/rook. This did simplify things a lot.

smithfarm commented:
@sebastian-philipp Bump of `src/seastar` submodule was intentioanal?

sebastian-philipp commented:
> @sebastian-philipp Bump of `src/seastar` submodule was intentional?

whoops! Glad we have a "Unmodified Submodules" check for that.

smithfarm commented on ceph.spec.in:
```diff
@@ -563,6 +563,7 @@ Summary:        Ceph Manager plugin for Rook-based orchestration
 %if 0%{?suse_version}
 Group:          System/Filesystems
 %endif
+BuildRequires:  python3-pip
```
this should be `python%{python3_pkgversion}-pip` I think

smithfarm commented on src/tools/setup-virtualenv.sh:
```diff
@@ -60,17 +60,17 @@ mkdir -p $DIR
 . $DIR/bin/activate
 
-if pip --help | grep -q disable-pip-version-check; then
+if python -m pip --help | grep -q disable-pip-version-check; then
```
At least on SUSE, `python` means Python 2 - is that intended?

jmolmo commented on src/pybind/mgr/rook/rook_cluster.py:
```diff
@@ -499,80 +480,82 @@ def add_osds(self, drive_group, all_hosts):
+
+                if block_devices:
+                    pd.devices = ccl.DevicesList(
+                        ccl.DevicesItem(name=d) for d in block_devices
```
ccl.DevicesItem(name=d.path) for d in block_devices

jmolmo replied:
@sebastian-philipp , this change is needed in order to make possible to create OSD's

jmolmo commented on src/pybind/mgr/rook/rook_cluster.py:
```diff
@@ -499,80 +480,82 @@ def add_osds(self, drive_group, all_hosts):
+                                current_node.devices = ccl.DevicesList()
+                            new_devices = list(set(block_devices) - set([d.name for d in current_node.devices]))
+                            current_node.devices.extend(
+                                ccl.DevicesItem(name=n) for n in new_devices
```
ccl.DevicesItem(name=n.path) for n in new_devices

jmolmo replied:
@sebastian-philipp , this change is needed in order to make possible to create OSD's

jmolmo commented:
It would be nice to have also the rook_client python files generated. In my view, we should rely in a fixed "version" of the python client included in the repo, and do not generate files each time we need them.... 
This forces to avoid problems when changes happens in Rook. if we have a set of client files generated for Rook 1.2.3... then we should only support this version of Rook. 
I think that the use of an automatic generated rook python client only will be possible when we have a complete integration test coverage of all the orchestrator API for Rook.

sebastian-philipp commented:
> It would be nice to have also the rook_client python files generated.

? Do you mean we should generate the files, or do you mean we should commit the generated files?

> In my view, we should rely in a fixed "version" of the python client included in the repo, and do not generate files each time we need them....

That's already the case!

> I think that the use of an automatic generated rook python client only will be possible when we have a complete integration test coverage of all the orchestrator API for Rook.

That's a different story!

sebastian-philipp commented:
@tchaikov @smithfarm , I'm failing hard to download `rook_client`:

1. https://github.com/rook/rook/pull/4204 is stalled, as the decision was to **not** include this into the main Rook git tree!
2. Having https://github.com/sebastian-philipp/rook-client-python within the Rook org in github is out of scope for Octopus
3. I need this for Octopus!
4. Downloading the files is not trivial. E.g. 
    ```
    [   73s] Collecting rook-client from git+https://github.com/sebastian-philipp/rook-client-python@rook-v1.2.2#egg=rook-client
    [   73s]   Cloning https://github.com/sebastian-philipp/rook-client-python (to revision rook-v1.2.2) to /tmp/pip-install-kvsoomnj/rook-client
    [   73s]   Error [Errno 2] No such file or directory: 'git': 'git' while executing command git clone -q https://github.com/sebastian-philipp/rook-client-python /tmp/pip-install-kvsoomnj/rook-client
    [   73s] Cannot find command 'git' - do you have 'git' installed and in your PATH?
    ```

Right now, I'd need merged this is as soon as possible. Like yesterday. Do you have any ideas?

jmolmo commented:
> > It would be nice to have also the rook_client python files generated.
> 
> ? Do you mean we should generate the files instead of just downloading them, or do you mean we should commit the generated files?

Yes. We should submit the generated files and use them until a decent way to install the rook python client will be available in the  target systems.

sebastian-philipp commented:
Still fighting with opensuse builds:
```
fatal: unable to access 'https://github.com/ceph/rook-client-python/': Could not resolve host: github.com
```

sebastian-philipp commented:
http://pulpito.ceph.com/swagner-2020-02-12_13:04:59-rados:mgr-wip-swagner-testing-distro-basic-smithi/

sebastian-philipp commented:
@liewegas this is ready to merge. The QA failures are irrelevant for this PR.

sebastian-philipp commented:
This adds a submodule, which means I cannot merge it.
